renovate the file upload

mawinter69 opened the pull request:
instead including the upload functionality in an iframe directly include it in the form. The file input has no name attribute so it will not be up uploaded when submitting.
When clicking the button the file will by uploaded asynchronously and a notification  bar is shown to the user with the result.

script and css is included via an adjunct is plain vanilla javascript.

requires #56  

New look:
<img width="434" alt="image" src="https://github.com/jenkinsci/sidebar-link-plugin/assets/17767050/0661e42c-cff1-421b-80e8-e21a5079e196">


<!-- Please describe your pull request here. -->

### Testing done

Manual testing. Go to Jenkins -> Manage -> Configure
Click on `Upload` under the `Add Link` button -> Warning is shown `No file chosen.`
Click on `Choose file` and select a file, click `Upload` -> Notification is shown where the file is found (The notification doesn't allow html) so the path is printed in single quotes. This notification is sticky so it will not disappear.
Click `Upload` again -> An error is shown that the file already exists.

Enter path to uploaded file in a Link and `Save` -> See the chosen image in the link.

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->
### Submitter checklist

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md in your own repository 
-->

damianszczepanik commented on src/main/resources/hudson/plugins/sidebar_link/Messages.properties:
```diff
@@ -1,4 +1,3 @@
-NoFile=No file uploaded.
-DupName=File with this name already exists in userContent.
-Back=Back
+uploaded=Uploaded image file; use {0} in Link Icon field above.
```
do we need to change it to lowercase ?

mawinter69 replied:
checkstyle complained that the methods used in java start with a capital letter

damianszczepanik replied:
it did not complain so far so I don't see reason why it starts now

mawinter69 replied:
It was spotbugs see https://ci.jenkins.io/job/Plugins/job/sidebar-link-plugin/job/PR-58/1/spotbugs/
Probably because I changed the Messages.properties

damianszczepanik replied:
can you revert those changes - I don't see value of them, just key renaming and loosing translations which were deleted

mawinter69 replied:
Done.
What might be done is move the ja translations from the deleted `startUpload_ja.properties`to the `config_ja.properties`. The texts are identical except for one where just the traliing `:` is removed.

damianszczepanik commented on src/main/resources/hudson/plugins/sidebar_link/SidebarLinkPlugin/upload.js:
```diff
@@ -0,0 +1,25 @@
+Behaviour.specify("#sidebar-link-upload-button", "sidebar-link-upload", 0, function(e) {
```
I'm not happy to merge this JS code, prefer to have ready to use library

damianszczepanik replied:
I believe this is not the only plugin that uploads the file and I believe there is one common reusable approach for Jenkins like for other controls

mawinter69 replied:
Usually files are uploaded during a form submit (e.g. when you have a build with a file parameter and you start a build).
But obviously this is not desired here as you want to upload the file and then be able to configure it without the need to press apply or save. Otherwise the existing approach with the iframe would not be here.
This iframe is not really the way to go in my eyes. After uploading you get a simple html snippet that contains no styling and you need to press back to get back to the place where you can upload files.
Of course you can include the file upload in every submit of the Jenkins config though that seems also not the right approach for me.

damianszczepanik replied:
@daniel-beck do we have reusable component for file upload instead of having this custom JavaScript code ?

daniel-beck replied:
Not that I'm aware of.

damianszczepanik replied:
how about some opensource library eg https://github.com/pqina/filepond

damianszczepanik replied:
Jenkins is using pure <input> field for uploading the pluign

daniel-beck replied:
Feel free to use whatever you want in your plugin.

For something provided by core, to be generally usable by plugins, it should probably come in the form of a Jelly taglib to abstract away the implementation and future-proof it. So that'll require a bit of work, and this hasn't really come up before AFAIK.

mawinter69 replied:
I think using such a library is overkill. And it will not prevent that you need some javascript to connect the file input field with the library.  Also for the upload you need to include a crumb, so in the end I doubt that when using the library you will get less javascript code.
And half of the code that we have now is related to notifying the user of the result.
Having a taglib might be useful, but I'm not aware of any other use case like the one here, where the upload should not be done as part of a form submission using the save or apply buttons.
If you want I can create a jelly taglib here in the plugin.

damianszczepanik replied:
> If you want I can create a jelly taglib here in the plugin.

Sounds like good plan @mawinter69 

I also found https://github.com/jenkinsci/jenkins/pull/7452

damianszczepanik commented:
This PR is conflicted with the master branch

mawinter69 commented:
Any plans to merge this?
On 2.387.3 the UI is broken now, the iframe causes scrolling.
<img width="809" alt="image" src="https://github.com/jenkinsci/sidebar-link-plugin/assets/17767050/9183d2e9-f9e3-4845-86d2-a059439fa065">
My change would fix this.

jonesbusy commented:
> Any plans to merge this?
> On 2.387.3 the UI is broken now, the iframe causes scrolling.

+1

damianszczepanik commented:
To be honest I don't feel that this change is the perfect way of fixing problem we have:
- this js file is hard to maintain
- js file has some hard coded not translatable labels
- I'm not able to validate custom css which are introduced here, prefer to reuse once that are already delivered by Jenkins
- and the most problematic is - is that JS code secure - I'm not able to validate it, you can't prove it

so the best for me is to have this component in Jenkins core and use it here as the reference.

Same problem is for plugin upload - ugly button is presented by Jenkins from a very long time

mawinter69 commented:
> To be honest I don't feel that this change is the perfect way of fixing problem we have:
> 
> * this js file is hard to maintain

Just because you don't feel comfortable with js doesn't mean it is hard to maintain. The script has a total of less than 30 lines. With a minimal experience in javascript and knowledge about some basic things that Jenkins provides for use in javascript (notificationBar, Behaviour)  it is easy to understand.

> * js file has some hard coded not translatable labels

I fixed that, there was one text that was hard coded, I made it translatable via jelly.
> * I'm not able to validate custom css which are introduced here, prefer to reuse once that are already delivered by Jenkins

I removed the css
> * and the most problematic is - is that JS code secure - I'm not able to validate it, you can't prove it

See my note above. With some basic knowledge in javascript one can easily see that the code is uncritical. If you want I can explain you for every line what it is doing so you can assure yourself that the code is secure. You can also ask some of the core maintainers if they are willing to verify if the code is secure. In Core basically all PR with changes in the UI are looked at from a security perspective e.g. if a change will introduce a XSS vulnerability.

> 
> so the best for me is to have this component in Jenkins core and use it here as the reference.

The [custom folder icon plugin ](https://github.com/jenkinsci/custom-folder-icon-plugin/blob/main/src/main/webapp/js/custom-icon-config.js) is also doing asynchronous file upload, but it also uses a js library named croppie in that method, that allows to resize/crop images. 
So I doubt this will ever become a core functionality.

> 
> Same problem is for plugin upload - ugly button is presented by Jenkins from a very long time

The upload of plugins in core is done by a form submit, as it is done for file parameters in jobs and in some other places and plugins. And there is no ugly button in recent Jenkins version.

You will need to decide if you want to keep your plugin broken as it is, accept this PR or find another way to fix it.
The feature to directly upload an image is just a convenience, in environments where you use configuration as code (be it via CasC plugin or some other way) you don't need it normally you would ensure that the required images are available. And as an admin you can also just write a simple freestyle job that runs on the controller and uploads a file to the userContent directory. 
So just completely removing the file upload would also be a valid option in my eyes, definitely better than doing nothing and have a broken UI.
